plugins/inputs/filecount/filecount.go

```diff
@@ -208,19 +228,30 @@ func (fc *FileCount) filter(file os.FileInfo) (bool, error) {
 }
 
 func (fc *FileCount) Gather(acc telegraf.Accumulator) error {
-	globDirs := fc.getDirs()
-	dirs, err := getCompiledDirs(globDirs)
-	if err != nil {
-		return err
+	if fc.globPaths == nil {
+		fc.initGlobPaths(acc)
 	}
 
-	for _, dir := range dirs {
-		fc.count(acc, dir, fc.Recursive)
+	for _, glob := range fc.globPaths {
+		for _, dir := range onlyDirectories(glob.GetRoots()) {
```

Comment:
> It seems like we should update glob.GetRoots() to always return a list of directories, this seems more in line with the function name.

I think it's better to rename the function. For filecount, our objects are directories, that's why we are discarding non-directory elements. But other plugins may have the need to recurse by themselves (and not use the globpath.Walk function) to find files and not (only) directories. In the case of patterns such as "a/b/c/d**", they have to examine both files and directories returned by GetRoots.

> Let's add documentation to the new functions in globpath as well.

Done.

Reviewing globpath code, there is inconsistency introduced by my code in Match function regarding the 'Stat' on matched files : 

- when !hasMeta : stat-ing the 'path' prevents returning a non-existent file or a file we cannot stat (is there another error case ?)
- when !HasSuperMeta : also stat-ing results of filepath.Glob . Is it necessary ? What does it prevent ?
- when hasSuperMeta : not stat-ing results of filepath.Walk (and not filtering errors ... which may also be necessary)

Reply:
> I think it's better to rename the function.

Okay, I guess I'm not sure what a good name is here though.  If you think there is a more appropriate one then use that otherwise we can leave as is.

> regarding the 'Stat' on matched files

It may be there there for non-existing files, but I don't think it is needed.  It doesn't make sense to worry too much about errors opening the files as these need to be handled when actually opening them, also it probably just makes it more difficult to report errors later.  I think it is best to remove the stat calls and just return all items.